examples/stable/mysql-csi-snapshot/mysql-csi-snapshot-blueprint.yaml

```diff
@@ -0,0 +1,44 @@
+apiVersion: cr.kanister.io/v1alpha1
+kind: Blueprint
+metadata:
+  name: mysql-csi-snapshot-bp
+actions:
+  backup:
+    outputArtifacts:
+      snapshotInfo:
+        keyValue:
+          name: "{{ .Phases.createCSISnapshot.Output.name }}"
+          pvc: "{{ .Phases.createCSISnapshot.Output.pvc }}"
+          namespace: "{{ .Phases.createCSISnapshot.Output.namespace }}"
+          restoreSize: "{{ .Phases.createCSISnapshot.Output.restoreSize }}"
```

Comment:
The restoreSize field from the VolumeSnapshot is allowed to be zero, if unknown. Even if it is set, it is supposed to be the minimum size required for restoring the actual data.

Depending on the CSI driver, this might therefore completely fail (if restoreSize is zero), or it is possible that the size of the resulting PVC will be just as much as is needed to restore the data. In that case, you have your data back, but probably can't do much with the new PVC unless you manually clone it into a bigger one.

Is it possible to either manually specify the restore size, or even track the original PVC's size somehow?

---

I also tested it with OpenEBS cstor - restoreSize remains zero, leading to an error:

> Warning  ActionSetFailed Action: restore  14s   Kanister Controller  Failed to execute phase: v1alpha1.Phase{Name:"restoreCSISnapshot", State:"pending", Output:map[string]interface {}(nil)}: PersistentVolumeClaim "demo-cstor-vol5-restored" is invalid: spec.resources[storage]: Invalid value: "0": must be greater than zero

---

References:

> restoreSize represents the complete size of the snapshot in bytes. In dynamic snapshot creation case, this field will be filled in by the CSI snapshotter sidecar with the "size_bytes" value returned from CSI "CreateSnapshot" gRPC call. For a pre-existing snapshot, this field will be filled with the "size_bytes" value returned from the CSI "ListSnapshots" gRPC call if the driver supports it. When restoring a volume from this snapshot, the size of the volume MUST NOT be smaller than the restoreSize if it is specified, otherwise the restoration will fail. If not specified, it indicates that the size is unknown.

Snapshot spec (protobuf message):

> // Information about a specific snapshot.
message Snapshot {
  // This is the complete size of the snapshot in bytes. The purpose of
  // this field is to give CO guidance on how much space is needed to
  // create a volume from this snapshot. The size of the volume MUST NOT
  // be less than the size of the source snapshot. This field is
  // OPTIONAL. If this field is not set, it indicates that this size is
  // unknown. The value of this field MUST NOT be negative and **a size of
  // zero means it is unspecified.**
  int64 size_bytes = 1;

Reply:
@oliverblaha We do keep track of original PVC's restoreSize using the `outputArtifacts`. The backup action of blueprint records the restoreSize as an outputArtifact which is further used as an inputArtifact in restore action. So when we create a restore actionset from backup actionset, the kanister controller picks up the original PVC's restoreSize and allocates the same amount of memory to the newly restored PVC. You can refer [this](file:///home/infracloud/kanister/docs/_build/html/templates.html#artifacts) section in the docs to know more about Artifacts in Kanister.

You can also manually tweak restoreSize field in blueprint before running Restore steps. Just need to edit `.actions.restore.phases[0].args.restoreSize` to a suitable value in a format like 5Gi, 2Gi, etc.

Reply:
@oliverblaha Can you share the blueprint you used to understand the issue?

Reply:
@shlokchaudhari9 I used the blueprint from this PR (just gave it a different name, that's all).
I believe that PVC does not have a `restoreSize` - so I assume you're referring to the snapshot's restoreSize field, am I correct? Unfortunately, that's the one that may be not set (= zero).

Reply:
Yes. I am referring to the VolumeSnapshot's `restoreSize` field. The one mentioned in the Status of VolumeSnapshot. 

As far as I understand, `restoreSize` is a size_bytes status field returned by the [Snapshot Controller](https://github.com/kubernetes-csi/external-snapshotter). We can't set the value of` restoreSize` field to 0 and check for failure, as there is no way to mention `restoreSize` from the [Spec](https://github.com/kubernetes-csi/external-snapshotter/blob/ea7834929a37af7af5b63a979001cdddc0372f4e/client/apis/volumesnapshot/v1/types.go#L75).

In the blueprint, there are two `restoreSize` fields. Both of which have different meanings:
1. In the output of create snapshot function -This is the  `restoreSize` from the Status of VolumeSnapshot
2. In the input of restore snapshot function - This is the storage request to be made in the new PVC. 

The value in the second `restoreSize` field MUST be greater than or equal to the outputted `restoreSize` from create snapshot function.

The error log above is correct, since it was produced for 2nd case (PVC restore)

> I also tested it with OpenEBS cstor - restoreSize remains zero, leading to an error
Could you maybe help me with a detailed description of this test scenario? Mainly along the following lines.
1. Original PVC manifest
2. Steps used to perform backup action
3. Steps used to perform restore action
I am still curious about how you encountered this error without updating the example Blueprint from this PR. Maybe there's an edge-case we are not handling.

Reply:
IIRC, the field is zero by default, and then gets updated by the snapshot controller with the actual number of bytes as soon es the snapshot is ready, _if_ the ListSnapshots gRPC-call is supported and _if_ the size is known. Otherwise, it remains zero.

So I guess what happens here is simply OpenEBS not updating it, which seems to be perfectly legal.
Then the field remains zero, which is what you get back from the snapshotter. That's of course not very useful when trying to restore ;)

Is it possible to get the size of the PVC instead?

Reply:
I see. Right now, we don't save the actual storage request the original PVC makes. We use the `restoreSize` from snapshot for the new PVC. So, the size of original PVC is something you might have to edit in the blueprint before trying to restore.
However, if this turns out to be a bug, one way to fix this would be by adding storage request size from original PVC in the `outputAritifacts` of create action.
But, before we jump to conclusions, I will definitely try out the OpenEBS cStor CSI driver for myself and update my findings here.

Reply:
Hey @oliverblaha I have created an issue regarding above error that you had https://github.com/kanisterio/kanister/issues/1195
Let's track the progress regarding this use-case on the issue, hereafter.

PS: Could you please update the Kubernetes Version/Provider in environment details there? Also, please feel free to edit or add any information in the description :)